rwengine/src/engine/Garage.cpp

```diff
@@ -0,0 +1,887 @@
+}
+
+void Garage::doOnStartClosingEvent() {
+    auto player = engine->getPlayer();
+    auto plyChar = player->getCharacter();
+    auto playerPosition = plyChar->getPosition();
+    auto playerVehicle = plyChar->getCurrentVehicle();
+    bool playerIsInVehicle = playerVehicle != nullptr;
+    RW_UNUSED(playerPosition);
+    RW_UNUSED(playerIsInVehicle);
+
+    switch (type) {
+        case Type::Mission:
+        case Type::CollectCars1:
+        case Type::CollectCars2: {
+            player->setInputEnabled(false);
+            break;
+        }
+
+        case Type::BombShop1:
+        case Type::BombShop2:
+        case Type::BombShop3:
+        case Type::Respray: {
+            player->setInputEnabled(false);
+            playerVehicle->setHandbraking(true);
+            break;
+        }
+
+        case Type::MissionForCarToComeOut: {
+            break;
+        }
+
+        case Type::Crusher: {
+            break;
+        }
+
+        case Type::MissionKeepCar: {
+            break;
+        }
+
+        case Type::MissionToOpenAndClose: {
+            break;
+        }
+
+        case Type::MissionForSpecificCar: {
+            break;
+        }
+
+        case Type::MissionKeepCarAndRemainClosed: {
+            break;
+        }
+
+        default: { break; }
+    }
+}
+
+void Garage::tick(float dt) {
+    if (!doorObject) {
+        return;
+    }
+    if (!active) {
+        return;
+    }
+
+    needsToUpdate = false;
+
+    switch (state) {
+        case State::Opened: {
+            if (shouldClose()) {
+                state = State::Closing;
+                doOnStartClosingEvent();
+            }
+
+            break;
+        }
+
+        case State::Closed: {
+            if (shouldOpen()) {
+                state = State::Opening;
+                doOnStartOpeningEvent();
+            }
+
+            break;
+        }
+
+        case State::Opening: {
+            if (shouldStopOpening()) {
+                state = State::Closing;
+            } else {
+                fraction += dt * step;
+
+                if (fraction >= 1.0f) {
+                    state = State::Opened;
+                    fraction = 1.f;
+                    doOnOpenEvent();
+                }
+
+                needsToUpdate = true;
+            }
+
+            break;
+        }
+
+        case State::Closing: {
+            if (shouldStopClosing()) {
+                state = State::Opening;
+            } else {
+                fraction -= dt * step;
+
+                if (fraction <= 0.f) {
+                    state = State::Closed;
+                    fraction = 0.f;
+                    doOnCloseEvent();
+                }
+
+                needsToUpdate = true;
+            }
+
+            break;
+        }
+
+        default: { break; }
+    }
+
+    if (needsToUpdate) {
+        updateDoor();
+    }
+}
+
+void Garage::updateDoor() {
+    if (swingType) {
+        doorObject->setRotation(glm::angleAxis(fraction * 1.57079632679f,
```

Comment:
Hardcoded value?

Reply:
yes, 90 degrees

Reply:
Maybe it would be clearer to put it as either "pi / 2", "2 * pi * 90 / 360" or "pi * 90 / 180"?

Reply:
You can use glm::half_pi. ;)